src/multivariate/dirichlet.jl

```diff
@@ -375,3 +375,56 @@ function fit_mle(::Type{<:Dirichlet}, P::AbstractMatrix{Float64},
     elogp = mean_logp(suffstats(Dirichlet, P, w))
     fit_dirichlet!(elogp, α; maxiter=maxiter, tol=tol, debug=debug)
 end
+
+## Differentiation
+function ChainRulesCore.frule((_, Δalpha), ::Type{DT}, alpha::AbstractVector{T}; check_args::Bool = true) where {T <: Real, DT <: Union{Dirichlet{T}, Dirichlet}}
+    d = DT(alpha; check_args=check_args)
+    Δalpha = ChainRulesCore.unthunk(Δalpha)
+    ∂alpha0 = sum(Δalpha)
+    digamma_alpha0 = SpecialFunctions.digamma(d.alpha0)
+    ∂lmnB = sum(Δalpha[i] * (SpecialFunctions.digamma(alpha[i]) - digamma_alpha0) for i in eachindex(alpha))
+    backing = (alpha=Δalpha, alpha0=∂alpha0, lmnB=∂lmnB)
+    t = ChainRulesCore.Tangent{typeof(d), NamedTuple{(:alpha, :alpha0, :lmnB), Tuple{typeof(alpha), typeof(d.alpha0), typeof(d.lmnB)}}}(backing)
+    return d, t
+end
+
+function ChainRulesCore.rrule(::Type{DT}, alpha::AbstractVector{T}; check_args::Bool = true) where {T <: Real, DT <: Union{Dirichlet{T}, Dirichlet}}
+    d = DT(alpha; check_args=check_args)
+    function dirichlet_pullback(d_dir)
+        d_dir = ChainRulesCore.unthunk(d_dir)
+        dalpha = d_dir.alpha .+ d_dir.alpha0 .+ d_dir.lmnB .* (SpecialFunctions.digamma.(alpha) .- SpecialFunctions.digamma.(d.alpha0))
+        return ChainRulesCore.NoTangent(), dalpha
+    end
+    return d, dirichlet_pullback
+end
+
+function ChainRulesCore.frule((_, Δd, Δx), ::typeof(_logpdf), d::Dirichlet, x::AbstractVector{<:Real})
+    lp = _logpdf(d, x)
+    if !insupport(d, x)
+        return (lp, zero(lp) + zero(eltype(Δx)) + zero(eltype(Δd.alpha)) + zero(eltype(Δd.lmnB)))
+    end
```

Comment:
It still feels somewhat unsatisfying to me since it is different from the computation in the common case `insupport(d, x)` and hence possibly in some special cases could lead to different types. It is also different from the usual approach taken in Distributions to ensure type stability.

Additionally, I think we should not return `0` but `NaN` (of the correct type) here: https://github.com/JuliaStats/Distributions.jl/pull/1490#issuecomment-1101305239

Generally, I think this branch could be removed if one uses `xlogy(Δd.alpha[i], max(zero(x[i]), x[i]))` (or some optimized version of it) below and returns `(lp, (insupport(d, x) ? ∂α_x + ∂l : oftype(∂α_x + ∂l, NaN))` (probably slightly simplified).

(In any case, `zero(eltype(Δd.lmnB))` could be replaced with `zero(Δd.lmnB)`.)

Reply:
yeah I read the argument for NaN values and it makes sense here